eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -339,55 +339,51 @@ proposalsRemoveIds gais ps =
       (roots, hierarchy) = foldl' removeEach (ps ^. pRootsL, ps ^. pGraphL) removedFromOMap
    in (checkInvariantAfterDeletion gais ps $ Proposals retainedOMap roots hierarchy, removedFromOMap)
   where
-    removeEach accum@(!roots, !hierarchy) gas =
-      withGovActionParent gas accum $ \forestL parent gpi ->
-        ( roots & forestL . prChildrenL %~ Set.delete gpi
-        , hierarchy
-            & forestL . pGraphNodesL %~ Map.delete gpi
-            & case parent of
-              SNothing -> id
-              SJust parentGpi ->
-                forestL . pGraphNodesL %~ Map.adjust (peChildrenL %~ Set.delete gpi) parentGpi
-        )
-
--- | Get all the descendents of an action-id from the @`Proposals`@ forest
-getAllDescendents ::
-  forall era.
-  Proposals era ->
-  GovActionId (EraCrypto era) ->
-  Set (GovActionId (EraCrypto era))
-getAllDescendents (Proposals omap _roots graph) gai = case OMap.lookup gai omap of
-  Nothing -> Set.empty
-  Just gas -> withGovActionParent gas Set.empty $ \graphL _ ->
-    Set.map unGovPurposeId . go graphL
-  where
-    go ::
-      forall p.
-      Lens' (GovRelation PGraph era) (PGraph (GovPurposeId p era)) ->
-      GovPurposeId p era ->
-      Set (GovPurposeId p era)
-    go graphL gpi =
-      case Map.lookup gpi $ graph ^. graphL . pGraphNodesL of
-        -- Impossible! getAllDescendents: GovPurposeId not found
-        Nothing -> assert False mempty
-        Just (PEdges _parent children) -> children <> foldMap (go graphL) children
-
--- | Remove the set of given action-ids with their descendents from the
--- @`Proposals`@ forest
-proposalsRemoveDescendentIds ::
+    removeEach accum@(!roots, !graph) gas =
+      withGovActionParent gas accum $ \govRelationL parent gpi ->
+        if parent == roots ^. govRelationL . prRootL
+          then
+            ( roots & govRelationL . prChildrenL %~ Set.delete gpi
+            , graph & govRelationL . pGraphNodesL %~ Map.delete gpi
+            )
+          else
+            ( roots
+            , graph
+                & govRelationL . pGraphNodesL %~ Map.delete gpi
+                & case parent of
+                  SNothing -> id
```

Comment:
Correct me if I'm wrong, but we _could_ have an `assert` here because this should not happen, but since this invariant is being checked in the check-function, we can ignore it is this branch?

Reply:
Yes, you are totally right. I've added the `assert` here, despite it being redundant, it serves as a good indicator to the reader that this is an impossible case.

Nice observation.